### ISSUE TITLE
url-encoded templates in notification payload are not replaced correctly

### ISSUE THREAD

**Issue by cdupont** [telefonicaid/fiware-orion#2903]
According to the [documentation](http://fiware-orion.readthedocs.io/en/master/user/forbidden_characters/), you can url-encode the payload. However if you do this, the templates (such as `${id}`) will not be replaced.

As an example:
I want to send this payload:
```
test: ${id}
```
I url-encode it:
```
$ urlencode "test: \${id}"
test%3A%20%24%7Bid%7D
```
The `\` is to prevent bash from interpreting the dollar sign.

I create the notification:
```
curl localhost:1026/v2/subscriptions -s -S --header 'Content-Type: application/json' --header 'Accept: application/json' -d @- <<EOF
{
  "description": "",
  "subject": {
    "entities": [
      {
        "id": "WS_UPPA_Sensor2",
        "type": "SensingDevice"
      }
    ],
    "condition": {
      "attrs": [
        "SM1"
      ]
    }
  },
  "notification": {
    "httpCustom": {
      "url": "http://localhost:10839",
      "headers": {
         "Content-Type": "application/json"
      },
      "method": "POST",      
      "payload": "test%3A%20%24%7Bid%7D"
    },
    "attrs": [
      "SM1"
    ]
  },
  "expires": "2040-01-01T14:00:00.00Z",
  "throttling": 5
}
EOF
```
Listening on the notification port I get:
```
$ ncat -l 10839 
test ${id}
```
The `${id}` is decoded but not interpreted by Orion.

PS. For url-encoding I used the alias proposed here:
https://unix.stackexchange.com/questions/159253/decoding-url-encoding-percent-encoding

**Comment by cdupont** [comment]
If I don't url-encode `${id}` it works though.

**Comment by fgalan** [comment]
Yes, that's the way it works. First, macro substitution. Next, url-decoding. Note that given that neither `$`, `{` or `}` are forbidden chars there is no need of encoding them.

However, I understand that your point is that your plan to apply the same encoding process at client side, no matter which characterse are in the payload field. Thus, makeing the encoding char-dependant is more complicated for your client. Is that correct?

**Comment by cdupont** [comment]
Yes, it's correct.
Is it a problem to swap "macro substitution" and "url-decoding" in Orion (start with url-decoding)?
I think it would be more correct/homogeneous because either we encode all the payload field, or none of it...
Encoding only a part of it is harder.
By the way, just for curiosity could you point me where this decoding is done in Orion? I could not find it...

**Comment by fgalan** [comment]
This is the point https://github.com/telefonicaid/fiware-orion/blob/0ee211472d6194a3d2123c0c5d92a09b87b2afbd/src/lib/ngsiNotify/Notifier.cpp#L238:

```
macroSubstitute(&payload, httpInfo.payload, ce);
char* pload = curl_unescape(payload.c_str(), payload.length());
```

Swapping both lines I think you would get your desired effect.

However, before consolidating a change like that is Orion, let's think a bit about the implications. Let's assume you have an attribute `A` which value is `%28foo%29` (which decoded is equal to `(foo)`). If you apply macro replacement first, then unescaping 

```
This is the value: ${A}
```
will result in:

```
This is the value: %28foo%29
```

which probably is not the result the user wants.

On the contraty, using the current approach the same

```
This is the value: ${A}
```
will result in:

```
This is the value: (foo)
```

which is probablty the one the user wants.

**Comment by cdupont** [comment]
So if I understand, both the *payload field* and the *attributes content* can be url-encoded?
In this case, I think that curl_unescape should be applied to both separately.
- upon receiving the payload, url-decode it.
- start decoding macros
- apply url-decode on each attributes that are found.

Another quirk to think about is that url-decode is not idempotent :(
So always decoding payloads might be dangerous (OK only in some rare cases).

Another thought is that it's strange to deal with URL-encoded strings internally in Orion.
I my opinion internally they should be decoded.
https://stackoverflow.com/questions/31825687/how-to-avoid-double-encoding-uri?answertab=votes#tab-top

Just my 2-cent :)

**Comment by fgalan** [comment]
Thank you very much for your valuable feedback. It would be very useful in the future if the encoding aspect takes relevance again (at the present moment I'm afraid they aren't a priority for the Orion development team...)

**Comment by cdupont** [comment]
Hello, I encountered this problem again.
I'm URL-encoding my payloads before submitting them to Orion, however the templates such as `${id}` will not be interpreted.

**Comment by fgalan** [comment]
> 
> 
> Hello, I encountered this problem again.
> I'm URL-encoding my payloads before submitting them to Orion, however the templates such as `${id}` will not be interpreted.

For reference, could you provide the exact request you are using, please? curl examples is the preferred way :) Thanks!

**Comment by cdupont** [comment]
Hi Firmin, the curl request in my first post works.

**Comment by kzangeli** [comment]
Quite clear what's happening here ...
The code that translates the "template" to the "actual" payload for the custom notification is looking for the exact chars '$', '{', and '}'.
The broker would have to also look for the url-encoded chains ($ == '%' + '2' + '4', etc) and treat those 3 chars as '$' (and so on) and do the same replacement. It's an easy fix.

We wouldn't want to force the users to never url-encode ... right?
Not that there's any real reason to url-encode the content of the payload body ...

At least as a workaround, try not encoding the `"payload": "test%3A%20%24%7Bid%7D"`

**Comment by cdupont** [comment]
Now as a workaround I URL-encode only forbidden characters. I need to do it because my payload is JSON, it has some `"` characters.

**Comment by fgalan** [comment]
@cdupont we have recently develop a funcionality to work "natively" with JSON in custom notification payload (see more details here: https://github.com/telefonicaid/fiware-orion/blob/master/doc/manuals/orion-api.md#json-payloads). Maybe it could be useful for your use case.

The new feature is already in master and it will be included in next Orion 3.8.0 release. At the present moment you can test it compiling from master or using `:latest` docker tag available.
